### PR TITLE
JDK-8272192 Shortcut String equality checks by checking equality of the value array

### PR DESCRIPTION
**Pull request by q2q-2q2**
Shortcut String equality checks by checking equality of the value array

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8272192](https://bugs.openjdk.java.net/browse/JDK-8272192): Shortcut String equality checks by checking equality of the value array

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5370/head:pull/5370` \
`$ git checkout pull/5370`

Update a local copy of the PR: \
`$ git checkout pull/5370` \
`$ git pull https://git.openjdk.java.net/jdk pull/5370/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5370`

View PR using the GUI difftool: \
`$ git pr show -t 5370`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5370.diff">https://git.openjdk.java.net/jdk/pull/5370.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/String.java [hunk 1]
```diff
@@ -1957,9 +1960,14 @@ public boolean contentEquals(CharSequence cs) {
      * @see  #codePoints()
      */
     public boolean equalsIgnoreCase(String anotherString) {
+        if (anotherString != null) {
+            return false;
```
Comment (liach) [inline src/java.base/share/classes/java/lang/String.java:1964]: This one is definitely wrong.

Comment (q2q-2q2) [reply src/java.base/share/classes/java/lang/String.java:1964]: Thank you

Comment (dholmes-ora) [reply src/java.base/share/classes/java/lang/String.java:1964]: So that begs the question as to what testing you did for this change?

**Comment by dholmes-ora** [comment]
It isn't obvious to me that this is necessarily correct - do we only ever share the full array? IIRC at one time substrings also shared the same array but with different a starting index and length.

If this is correct then it needs some supporting performance data, but as these methods are also VM intrinsics I doubt you will see much difference as only interpreted code will be affected by your change. But I'll leave it to the performance experts to comment - paging @cl4es :)

Cheers,
David

**Review by cl4es — COMMENTED** [review]
David is right that many of these methods have VM intrinsics that'd have to be fixed to see much benefit of this. 

I can imagine some of these to be profitable in edge cases with string de-duplication enabled. Otherwise the backing arrays will typically never be shared and the identity test always be false (`""` is an exception to that rule that interestingly is unlikely to see any win from short-cutting). While the gain from a correct shortcut could be really large on a synthetic micro-benchmark, I still suspect the added test might be more costly in aggregate on the fabled "real application".

With compact strings it's easy to construct strings with different coders but the same `byte[]`, and string de-duplication could be implemented so that you end up sharing the same array between two semantically different strings (I suspect this is the case, but I might be wrong). This means there's a possible correctness issue in some places where you're shortcutting before checking either length or coder. 

All things considered I don't think this is worthwhile.

**Pull request closed without merging**